lock-api/src/test/java/com/palantir/lock/client/CoalescingTransactionStarterTest.java

```diff
@@ -0,0 +1,201 @@
+
+        assertThat(CoalescingTransactionStarter.split(batchedResponse))
+                .hasSize(1)
+                .allSatisfy(startTxnResponse -> assertDerivableFromBatchedResponse(startTxnResponse, batchedResponse));
+    }
+
+    @Test
+    public void splitShouldYieldCorrectStartTransactionResponses_multipleTransactions() {
+        StartTransactionResponseV4 batchedResponse = getStartTransactionResponse(10, 5);
+
+        List<StartIdentifiedAtlasDbTransactionResponse> responses =
+                CoalescingTransactionStarter.split(batchedResponse);
+
+        assertThatStartTransactionResponsesAreUnique(responses);
+        assertThat(responses)
+                .hasSize(5)
+                .allSatisfy(startTxnResponse -> assertDerivableFromBatchedResponse(startTxnResponse, batchedResponse));
+    }
+
+    @Test
+    public void shouldDeriveStartTransactionResponseFromBatchedResponse_singleTransaction() {
+        StartTransactionResponseV4 startTransactionResponse = getStartTransactionResponse(12, 1);
+
+        when(lockLeaseService.startTransactions(1)).thenReturn(startTransactionResponse);
+        StartIdentifiedAtlasDbTransactionResponse response =
+                coalescingTransactionStarter.startIdentifiedAtlasDbTransaction();
+
+        assertDerivableFromBatchedResponse(response, startTransactionResponse);
+    }
+
+    @Test
+    public void shouldDeriveStartTransactionResponseFromBatchedResponse_multipleTransactions() {
+        StartTransactionResponseV4 batchResponse = getStartTransactionResponse(40, 3);
+        when(lockLeaseService.startTransactions(3))
+                .thenReturn(batchResponse);
+
+        List<StartIdentifiedAtlasDbTransactionResponse> responses = requestBatches(3);
+        assertThatStartTransactionResponsesAreUnique(responses);
+        assertThat(responses)
+                .hasSize(3)
+                .allSatisfy(startTxnResponse -> assertDerivableFromBatchedResponse(startTxnResponse, batchResponse));
+    }
+
+    @Test
+    public void shouldCallTimelockMultipleTimesUntilCollectsAllRequiredTimestamps() {
+        when(lockLeaseService.startTransactions(anyInt()))
+                .thenReturn(getStartTransactionResponse(40, 2))
+                .thenReturn(getStartTransactionResponse(100, 1));
+
+        requestBatches(3);
+        verify(lockLeaseService).startTransactions(3);
+        verify(lockLeaseService).startTransactions(1);
+
+    }
+
+    private List<StartIdentifiedAtlasDbTransactionResponse> requestBatches(int size) {
+        List<BatchElement<Void, StartIdentifiedAtlasDbTransactionResponse>> elements = IntStream.range(0, size)
+                .mapToObj(unused -> ImmutableTestBatchElement.builder()
+                        .argument(null)
+                        .result(SettableFuture.create())
+                        .build())
+                .collect(toList());
+        CoalescingTransactionStarter.consumer(lockLeaseService).accept(elements);
```

Comment:
this feels quite iffy, wondering whether there is a lighter approach to "mocking"

Reply:
it feels like you want to control the batch size given to the autobatcher, or that the consumer interface needs to be clearer i.e.:

`Function<List<Request<T>, Map<Request<T>, R>>`

I dunno, just feels off.

Reply:
yeah, it seems like too much tooling for a test; but wanted to cover as much code as possible without testing `AutoBatcher` - so consumer has exactly the same signature required by the `AutoBatcher`. Used a similar approach with `RequestBatchingTimetampServiceTest`